src/elliptic/curves/secp256_r1.rs

```diff
@@ -0,0 +1,972 @@
+        let pk = base_point.scalar_mul(&random_scalar.get_element());
+        Secp256r1Point {
+            purpose: "random_point",
+            ge: pk.get_element(),
+        }
+    }
+
+    pub fn base_point2() -> Secp256r1Point {
+        Secp256r1Point::from_coor(
+            &BigInt::from(BASE_POINT2_X.as_ref()),
+            &BigInt::from(BASE_POINT2_Y.as_ref()),
+        )
+    }
+
+    fn to_negative(&self) -> Result<Self, ErrorKey> {
+        let order = BigInt::from(&FIELD_MODULO[..]);
+        let x = self.x_coor().unwrap();
+        let y = self.y_coor().unwrap();
+        let minus_y = BigInt::mod_sub(&order, &y, &order);
+
+        let x_vec = BigInt::to_vec(&x);
+        let y_vec = BigInt::to_vec(&minus_y);
+
+        let mut template_x = vec![0; 32 - x_vec.len()];
+        template_x.extend_from_slice(&x_vec);
+        let mut x_vec = template_x;
+
+        let mut template_y = vec![0; 32 - y_vec.len()];
+        template_y.extend_from_slice(&y_vec);
+        let y_vec = template_y;
+
+        x_vec.extend_from_slice(&y_vec);
+
+        Self::from_bytes(&x_vec)
+    }
+}
+
+impl Zeroize for Secp256r1Scalar {
+    fn zeroize(&mut self) {
+        unsafe { ptr::write_volatile(self, FE::zero()) };
+        atomic::fence(atomic::Ordering::SeqCst);
+        atomic::compiler_fence(atomic::Ordering::SeqCst);
+    }
+}
+
+impl ECScalar<Seed> for Secp256r1Scalar {
+    fn new_random() -> Secp256r1Scalar {
+        let mut arr = [0u8; SECRET_KEY_SIZE];
+        thread_rng().fill(&mut arr[..]);
+        Secp256r1Scalar {
+            purpose: "random",
+            fe: Seed::from_bytes(
+                &P256,
+                untrusted::Input::from(&arr.to_vec()),
+                cpu::features(), // TODO: remove cpu to be encapsulated on the seed
+            )
+            .unwrap(), // TODO: handle unwrap
+        }
+    }
+
+    fn zero() -> Self {
+        Secp256r1Scalar {
+            purpose: "zero",
+            fe: Seed::zero(&P256, cpu::features()),
+        }
+    }
+
+    fn get_element(&self) -> Seed {
+        self.fe
+    }
+
+    fn set_element(&mut self, element: Seed) {
+        self.fe = element
+    }
+
+    fn from(n: &BigInt) -> Secp256r1Scalar {
+        let curve_order = Self::q();
+        let n_reduced = n.mod_floor(&curve_order);
+        let mut v = BigInt::to_vec(&n_reduced);
+
+        if v.len() < SECRET_KEY_SIZE {
+            let mut template = vec![0; SECRET_KEY_SIZE - v.len()];
+            template.extend_from_slice(&v);
+            v = template;
+        }
+
+        Secp256r1Scalar {
+            purpose: "from_big_int",
+            fe: Seed::from_bytes(&P256, untrusted::Input::from(&v), cpu::features()).unwrap(),
+        }
+    }
+
+    fn to_big_int(&self) -> BigInt {
+        BigInt::from(self.fe.bytes_less_safe())
+    }
+
+    fn q() -> BigInt {
+        BigInt::from(CURVE_ORDER.as_ref())
+    }
+
+    fn add(&self, other: &Seed) -> Secp256r1Scalar {
```

Comment:
Ring is not providing a way to do add/mul/sub/neg?

Reply:
Not for a field element.
I followed the same pattern as used in the Secp256k1 source, see https://github.com/KZen-networks/curv/blob/master/src/elliptic/curves/secp256_k1.rs#L160.